File: flutter_application_1/lib/models/complaint.dart
enum ComplaintPriority { low, medium, high, critical }

enum ComplaintStatus { open, inProgress, resolved, closed }

ComplaintPriority complaintPriorityFromString(String? value) {
  switch (value) {
    case 'LOW':
      return ComplaintPriority.low;
    case 'HIGH':
      return ComplaintPriority.high;
    case 'CRITICAL':
      return ComplaintPriority.critical;
    case 'MEDIUM':
    default:
      return ComplaintPriority.medium;
  }
}

String complaintPriorityToString(ComplaintPriority priority) {
  switch (priority) {
    case ComplaintPriority.low:
      return 'LOW';
    case ComplaintPriority.medium:
      return 'MEDIUM';
    case ComplaintPriority.high:
      return 'HIGH';
    case ComplaintPriority.critical:
      return 'CRITICAL';
  }
}

ComplaintStatus complaintStatusFromString(String? value) {
  switch (value) {
    case 'IN_PROGRESS':
      return ComplaintStatus.inProgress;
    case 'RESOLVED':
      return ComplaintStatus.resolved;
    case 'CLOSED':
      return ComplaintStatus.closed;
    case 'OPEN':
    default:
      return ComplaintStatus.open;
  }
}

class Complaint {
  final int? id;
  final String title;
  final String description;
  final double latitude;
  final double longitude;
  final ComplaintPriority priority;
  final ComplaintStatus status;
  final String? imageUrl;
  final DateTime? createdAt;

  Complaint({
    this.id,
    required this.title,
    required this.description,
    required this.latitude,
    required this.longitude,
    this.priority = ComplaintPriority.medium,
    this.status = ComplaintStatus.open,
    this.imageUrl,
    this.createdAt,
  });

  factory Complaint.fromJson(Map<String, dynamic> json) {
    return Complaint(
      id: json['id'] as int?,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0,
      priority: complaintPriorityFromString(json['priority'] as String?),
      status: complaintStatusFromString(json['status'] as String?),
      imageUrl: json['imageUrl'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toCreateJson(Map<String, dynamic> userRef) {
    return {
      'title': title,
      'description': description,
      'latitude': latitude,
      'longitude': longitude,
      'priority': complaintPriorityToString(priority),
      'user': userRef,
    };
  }
}

File: ecolink-back/src/main/java/com/ecolink/model/Complaint.java
package com.ecolink.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "complaints", indexes = {
    @Index(name = "idx_complaint_userId", columnList = "user_id"),
    @Index(name = "idx_complaint_priority", columnList = "priority"),
    @Index(name = "idx_complaint_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Complaint {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private Double latitude;

    @Column(nullable = false)
    private Double longitude;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Priority priority;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ComplaintStatus status;

    @Column(length = 255)
    private String imageUrl;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public enum Priority {
        LOW, MEDIUM, HIGH, CRITICAL
    }

    public enum ComplaintStatus {
        OPEN, IN_PROGRESS, RESOLVED, CLOSED
    }

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        if (this.createdAt == null) this.createdAt = now;
        if (this.updatedAt == null) this.updatedAt = now;
        if (this.priority == null)  this.priority  = Priority.MEDIUM;
        if (this.status == null)    this.status    = ComplaintStatus.OPEN;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
